src/aiodynamo/expressions.py

```diff
@@ -0,0 +1,537 @@
+
+    def encode_name(self, name: Union[str, int]) -> str:
+        return self._encode(
+            "#n", name, self.names, self.names_gen, self.names_cache, name
+        )
+
+    def encode_value(self, value: Any) -> Optional[str]:
+        value = low_level_serialize(value)
+        if value:
+            return self._encode(
+                ":v",
+                {value[0]: value[1]},
+                self.values,
+                self.values_gen,
+                self.values_cache,
+                value,
+            )
+        return None
+
+    def encode_path(self, path: KeyPath) -> str:
+        bits = [self.encode_name(path[0])]
+        for part in path[1:]:
+            if isinstance(part, int):
+                bits.append(f"[{part}]")
+            else:
+                bits.append(f".{self.encode_name(part)}")
+        return "".join(bits)
+
+    def get_expression_names(self) -> Dict[str, str]:
+        return self.names
+
+    def get_expression_values(self) -> Dict[str, Dict[str, Any]]:
+        return {key: value for key, value in self.values.items()}
+
+    def _encode(
+        self,
+        prefix: str,
+        thing: Any,
+        data: Dict[str, Any],
+        index_gen: Iterator[int],
+        cache: _ParametersCache,
+        cache_key: Any,
+    ) -> str:
+        try:
+            return cache[cache_key]
+
+        except KeyError:
+            can_cache = True
+        except TypeError:
+            can_cache = False
+        encoded = f"{prefix}{next(index_gen)}"
+        data[encoded] = thing
+        if can_cache:
+            cache[cache_key] = encoded
+        return encoded
+
+
+@dataclass(frozen=True)
+class HashAndRangeKeyCondition(KeyCondition):
+    hash_key: HashKey
+    range_key_condition: Condition
+
+    def encode(self, params: Parameters) -> str:
+        return f"{self.hash_key.encode(params)} AND {self.range_key_condition.encode(params)}"
+
+
+class Condition(metaclass=abc.ABCMeta):
+    def __and__(self, other: Condition) -> Condition:
+        return AndCondition(self, other)
+
+    def __or__(self, other: Condition) -> Condition:
+        return OrCondition(self, other)
+
+    def __invert__(self) -> Condition:
+        return NotCondition(self)
+
+    @abc.abstractmethod
+    def encode(self, params: Parameters) -> str:
+        pass
+
+
+@dataclass(frozen=True)
+class NotCondition(Condition):
+    base: Condition
+
+    def encode(self, params: Parameters) -> str:
+        return f"NOT {self.base.encode(params)}"
+
+
+@dataclass(frozen=True)
+class AndCondition(Condition):
+    lhs: Condition
+    rhs: Condition
+
+    def encode(self, params: Parameters) -> str:
+        return f"({self.lhs.encode(params)} AND {self.rhs.encode(params)})"
+
+
+@dataclass(frozen=True)
+class OrCondition(Condition):
+    lhs: Condition
+    rhs: Condition
+
+    def encode(self, params: Parameters) -> str:
+        return f"({self.lhs.encode(params)} OR {self.rhs.encode(params)})"
```

Comment:
I have some vague recollection that dynamo had weird requirements on what must be and what must not be parenthesized... Consider adding tests (against real dynamo) for cases like:

`x&y&z` if that should be `x AND y AND z` vs. `(x AND y) AND z` vs. `x AND (y AND z)`...
`x&y|z`, `x|y&z`, `x&!z`, etc..

Reply:
according to the [docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.OperatorsAndFunctions.html#Expressions.OperatorsAndFunctions.Precedence), you can parenthesize any condition. [boto3](https://github.com/boto/boto3/blob/develop/boto3/dynamodb/conditions.py#L224) also does the parens like this for AND/OR

Reply:
Hmmm maybe they changed something.
In the past, dynamo was very pedantic.. it would only accept "optimally parenthesised" queries, meaning, for example, no redundant parentheses were allowed.